### PR TITLE
Add support for Multiple VolumeSnapshotClasses

### PR DESCRIPTION
**Pull request by anshulahuja98**
Fix: [#5750](https://github.com/vmware-tanzu/velero/issues/5750)

Design: https://github.com/vmware-tanzu/velero/pull/5774

### CODE REVIEW

File: internal/util/util.go [hunk 1]
```diff
@@ -114,13 +114,73 @@ func IsPVCDefaultToFSBackup(pvcNamespace, pvcName string, podClient corev1client
...
+// GetVolumeSnapshotClassFromAnnotationsForStorageClass returns a VolumeSnapshotClass for the supplied volume provisioner/ driver name from the annotation of the backup
+func GetVolumeSnapshotClassFromBackupAnnotationsForDriver(backup *velerov1api.Backup, provisioner string, snapshotClasses *snapshotv1api.VolumeSnapshotClassList) (*snapshotv1api.VolumeSnapshotClass, error) {
+	annotationKey := fmt.Sprintf("%s/%s", VolumeSnapshotClassDriverBackupAnnotationPrefix, strings.ToLower(provisioner))
+	snapshotClassName, ok := backup.ObjectMeta.Annotations[annotationKey]
+	if !ok {
+		return nil, nil
+	}
+	for _, sc := range snapshotClasses.Items {
+		if strings.EqualFold(snapshotClassName, sc.ObjectMeta.Name) {
+			return &sc, nil
```
Comment (Shashank1306s) [inline internal/util/util.go:176]: Can we add similar check as we do for validating snapshot class driver and provisioner in the above function (GetVolumeSnapshotClassFromPVCAnnotationsForDriver)?

**Review by Shashank1306s — COMMENTED** [review]
Added comments in internal/util/util.go

File: internal/util/util.go (hunk 1 above)
Comment (anshulahuja98) [reply internal/util/util.go:176]: Makes sense. Will prevent user caused errors.

File: internal/util/util.go [hunk 2]
```diff
@@ -114,13 +114,73 @@ func IsPVCDefaultToFSBackup(pvcNamespace, pvcName string, podClient corev1client
...
-func GetVolumeSnapshotClassForStorageClass(provisioner string, snapshotClient snapshotter.SnapshotV1Interface) (*snapshotv1api.VolumeSnapshotClass, error) {
+func GetVolumeSnapshotClass(pvc *corev1api.PersistentVolumeClaim, provisioner string, backup *velerov1api.Backup, log logrus.FieldLogger, snapshotClient snapshotter.SnapshotV1Interface) (*snapshotv1api.VolumeSnapshotClass, error) {
 	snapshotClasses, err := snapshotClient.VolumeSnapshotClasses().List(context.TODO(), metav1.ListOptions{})
 	if err != nil {
 		return nil, errors.Wrap(err, "error listing volumesnapshot classes")
 	}
+	// If a snapshot class is sent for provider in PVC annotations, use that
+	snapshotClass, err := GetVolumeSnapshotClassFromPVCAnnotationsForDriver(pvc, provisioner, snapshotClasses)
+	if err != nil {
+		// log error
```
[the remark this answers is not in the document]
Comment (anshulahuja98) [reply internal/util/util.go:125]: ack

File: internal/util/labels_annotations.go [hunk 3]
```diff
@@ -17,14 +17,16 @@ limitations under the License.
...
-	VolumeSnapshotClassSelectorLabel = "velero.io/csi-volumesnapshot-class"
+	VolumeSnapshotLabel                             = "velero.io/volume-snapshot-name"
+	VolumeSnapshotHandleAnnotation                  = "velero.io/csi-volumesnapshot-handle"
+	VolumeSnapshotRestoreSize                       = "velero.io/vsi-volumesnapshot-restore-size"
+	CSIDriverNameAnnotation                         = "velero.io/csi-driver-name"
+	CSIDeleteSnapshotSecretName                     = "velero.io/csi-deletesnapshotsecret-name"
+	CSIDeleteSnapshotSecretNamespace                = "velero.io/csi-deletesnapshotsecret-namespace"
+	CSIVSCDeletionPolicy                            = "velero.io/csi-vsc-deletion-policy"
+	VolumeSnapshotClassSelectorLabel                = "velero.io/csi-volumesnapshot-class"
+	VolumeSnapshotClassDriverBackupAnnotationPrefix = "velero.io/csi-volumesnapshot-class"
```
Comment (blackpiglet) [inline internal/util/labels_annotations.go:28]: Looks like `VolumeSnapshotClassDriverBackupAnnotationPrefix` and `VolumeSnapshotClassDriverPVCAnnotation` are duplicated with `VolumeSnapshotClassSelectorLabel`.

Comment (anshulahuja98) [reply internal/util/labels_annotations.go:28]: Actually all 3 variables including VolumeSnapshotClassSelectorLabel  have the same string
I kept that intentionally same to make user experience easier
While on the other hand - all are separate 3 variables such that it is easy to change in code than reusing the variables.

Do you have any suggestion for keeping the values different?

**Comment by anshulahuja98** [comment]
CC @shubham-pampattiwar for review

**Review by sseago — COMMENTED** [review]
Added some comments. We can't merge this until we get the design PR approved and merged, though.

File: internal/util/util.go [hunk 4]
```diff
@@ -114,13 +114,75 @@ func IsPVCDefaultToFSBackup(pvcNamespace, pvcName string, podClient corev1client
 
 	return false, nil
 }
-
-// GetVolumeSnapshotClassForStorageClass returns a VolumeSnapshotClass for the supplied volume provisioner/ driver name.
-func GetVolumeSnapshotClassForStorageClass(provisioner string, snapshotClient snapshotter.SnapshotV1Interface) (*snapshotv1api.VolumeSnapshotClass, error) {
+func GetVolumeSnapshotClass(pvc *corev1api.PersistentVolumeClaim, provisioner string, backup *velerov1api.Backup, log logrus.FieldLogger, snapshotClient snapshotter.SnapshotV1Interface) (*snapshotv1api.VolumeSnapshotClass, error) {
```
[the remark this answers is not in the document]
Comment (anshulahuja98) [reply internal/util/util.go]: Done

File: internal/util/util.go [hunk 5]
```diff
@@ -114,13 +114,75 @@ func IsPVCDefaultToFSBackup(pvcNamespace, pvcName string, podClient corev1client
...
+			if !strings.EqualFold(sc.Driver, provisioner) {
+				return nil, errors.Errorf("Incorrect volumesnapshotclass, snapshot class %s is not for driver %s", sc.ObjectMeta.Name, provisioner)
+			}
+			return &sc, nil
+		}
+	}
+	return nil, errors.Errorf("No CSI VolumeSnapshotClass found with name %s for provisioner %s for PVC %s", snapshotClassName, provisioner, pvc.Name)
+}
+
+// GetVolumeSnapshotClassFromAnnotationsForStorageClass returns a VolumeSnapshotClass for the supplied volume provisioner/ driver name from the annotation of the backup
```
[the remark this answers is not in the document]
Comment (anshulahuja98) [reply internal/util/util.go]: Done

File: internal/backup/pvc_action.go [hunk 6]
```diff
@@ -105,7 +105,7 @@ func (p *PVCBackupItemAction) Execute(item runtime.Unstructured, backup *velerov
 		return nil, nil, errors.Wrap(err, "error getting storage class")
 	}
 	p.Log.Debugf("Fetching volumesnapshot class for %s", storageClass.Provisioner)
-	snapshotClass, err := util.GetVolumeSnapshotClassForStorageClass(storageClass.Provisioner, snapshotClient.SnapshotV1())
+	snapshotClass, err := util.GetVolumeSnapshotClass(&pvc, storageClass.Provisioner, backup, p.Log, snapshotClient.SnapshotV1())
```
[the remark this answers is not in the document]
Comment (anshulahuja98) [reply internal/backup/pvc_action.go]: Done

**Comment by anshulahuja98** [comment]
> Added some comments. We can't merge this until we get the design PR approved and merged, though.

Resolved all comments and merge conflicts
@sseago  Can you please request someone from the maintainers team to review this PR/ approve design PR. 
At this point I have requested too many times in the community meetings without receiving traction

**Comment by shubham-pampattiwar** [comment]
Lets get the design merged in first.

**Comment by anshulahuja98** [comment]
The design PR has been approved by 2 maintainers, would be useful if someone can merge it and also meanwhile review/ approve this current PR.

**Comment by anshulahuja98** [comment]
I am not sure how this got closed, must have done by mistake
@blackpiglet / @sseago / @shubham-pampattiwar  can you please approve this.

**Comment by anshulahuja98** [comment]
Design PR is merged

**Review by sseago — APPROVED** [review]

**Comment by anshulahuja98** [comment]
@blackpiglet  / @shubham-pampattiwar  can you please do final signoff and get this merged.

**Review by blackpiglet — APPROVED** [review]

**Comment by anshulahuja98** [comment]
@reasonerjt  can we get this merged as well?

**Pull request merged**